### ISSUE TITLE
VMWare update does not work in python3 venv

### ISSUE THREAD

**Issue by kdelee** [ansible/awx#4156]
##### ISSUE TYPE
 - Bug Report

##### SUMMARY

VMware updates do not work when inventory update is run in python3.

##### ENVIRONMENT
* AWX version: 6.0.0
* AWX install method: RPM
* Ansible version:  2.8
* Operating System: RHEL8
* Web Browser: n/a

##### STEPS TO REPRODUCE

In a python3 ansible virtualenv, run VMWare inventory update

On RHEL8, AWX 6.0.0 uses a python3 venv by default.

##### EXPECTED RESULTS

Works

##### ACTUAL RESULTS

```
Traceback (most recent call last):   File
"/var/lib/awx/venv/awx/lib64/python3.6/site-
packages/awx/plugins/inventory/vmware_inventory.py", line 772, in <module>
print(VMWareInventory().show())   File
"/var/lib/awx/venv/awx/lib64/python3.6/site-
packages/awx/plugins/inventory/vmware_inventory.py", line 145, in __init__
self.do_api_calls_update_cache()   File
"/var/lib/awx/venv/awx/lib64/python3.6/site-
packages/awx/plugins/inventory/vmware_inventory.py", line 184, in
do_api_calls_update_cache     self.inventory =
self.instances_to_inventory(self.get_instances())   File
"/var/lib/awx/venv/awx/lib64/python3.6/site-
packages/awx/plugins/inventory/vmware_inventory.py", line 351, in get_instances
return self._get_instances(kwargs)   File
"/var/lib/awx/venv/awx/lib64/python3.6/site-
packages/awx/plugins/inventory/vmware_inventory.py", line 393, in
_get_instances     for instance in sorted(instances): TypeError: '<' not
supported between instances of 'vim.VirtualMachine' and 'vim.VirtualMachine'

  File "/usr/lib/python3.6/site-packages/ansible/inventory/manager.py", line 267, in parse_source
    plugin.parse(self._inventory, self._loader, source, cache=cache)
  File "/usr/lib/python3.6/site-packages/ansible/plugins/inventory/script.py", line 161, in parse
    raise AnsibleParserError(to_native(e))
```

##### ADDITIONAL INFORMATION

Upstream bug: https://github.com/ansible/ansible/issues/46727

@abadger says he could provide some guidance to fix PR that was put up but not merged because it does not work in python2 (needs to work for both python2 and 3)

Basically we either need to fix the script we vendor or update to plugin to get past this.

**Comment by ryanpetrello** [comment]
@kdelee it seems like this change (which merged):

https://github.com/ansible/ansible/pull/47538

...might resolve the issue for us in the next Ansible 2.8.x.

Is that your understanding?

**Comment by kdelee** [comment]
:thinking: might .... this comment that was from after merge is slightly concerning https://github.com/ansible/ansible/pull/47538#issuecomment-505660986

Also looks like that went into `devel` a.k.a. 2.9, would need to get backported to 2.8 .... and potentially update our vendored copy of it? Feeling confused, will come back to this tomorrow.

**Comment by kdelee** [comment]
So refreshed my mind about this. Turns out we still use the vendored script for vmware. 

Plopped updated version of the script from ansible@devel https://raw.githubusercontent.com/ansible/ansible/devel/contrib/inventory/vmware_inventory.py into `/var/lib/awx/venv/awx/lib/python3.6/site-packages/awx/plugins/inventory/vmware_inventory.py` on a box with RPM install of tower and default ansible venv is python3 (RHEL8)

Was confused for a bit because although file lives in
`/var/lib/awx/venv/awx/lib/python3.6/site-packages/awx/plugins/inventory/vmware_inventory.py`
it is actually executed by `/var/lib/awx/venv/ansible/bin/ansible-inventory`, which uses python2 or python3 depending on system python


This appears to work :tada: 

still TODO: test same script on box that uses python2

**Comment by kdelee** [comment]
@ryanpetrello I hand-vendored the devel version of the vmware script https://raw.githubusercontent.com/ansible/ansible/devel/contrib/inventory/vmware_inventory.py onto a RHEL7 box where python2 is in the `ansible` venv....I was able to run an inventory update against our test VMware simulator...I cannot garuntee that this exercises all code paths, but FWIW seems to work.

Recommended path forward is to vendor the new script into `awx` and test it some more in python2 and python3 scenarios. Maybe we can find another VMWare inventory source to hit

**Comment by sankat447** [comment]
I updated my python script with https://raw.githubusercontent.com/ansible/ansible/devel/contrib/inventory/vmware_inventory.py at location /var/lib/awx/venv/awx/lib/python3.6/site-packages/awx/plugins/inventory/vmware_inventory.py script seems to work but Tower inventory source sync throw error.
1.599 INFO     Updating inventory 4: Platform Inventory
    2.011 DEBUG    Using base command: python /usr/bin/ansible-inventory -i /var/lib/awx/venv/awx/lib64/python3.6/site-packages/awx/plugins/inventory/vmware_inventory.py --playbook-dir /var/lib/awx/venv/awx/lib64/python3.6/site-packages/awx/plugins/inventory -vvvvv
    2.011 INFO     Reading Ansible inventory source: /var/lib/awx/venv/awx/lib64/python3.6/site-packages/awx/plugins/inventory/vmware_inventory.py
    2.012 INFO     Using VIRTUAL_ENV: /var/lib/awx/venv/ansible
    2.013 INFO     Using PATH: /var/lib/awx/venv/ansible/bin:/var/lib/awx/venv/awx/bin:/var/lib/awx/venv/awx/bin:/var/lib/awx/venv/awx/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin
    2.013 INFO     Using PYTHONPATH: /var/lib/awx/venv/ansible/lib/python3.6/site-packages:
    2.020 DEBUG    Using private credential data in '/tmp/awx_137_h02vftbb'.
    2.022 DEBUG    Using fresh temporary directory '/tmp/awx_proot_3n6llm0k' for isolation.
    2.022 DEBUG    Running from `/var/lib/awx/venv/awx/lib/python3.6/site-packages/awx/plugins/inventory` working directory.
   **14.667 ERROR    ansible-inventory 2.8.2
   14.667 ERROR      config file = /etc/ansible/ansible.cfg
   14.667 ERROR      configured module search path = ['/var/lib/awx/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
   14.667 ERROR      ansible python module location = /usr/lib/python3.6/site-packages/ansible
   14.667 ERROR      executable location = /usr/bin/ansible-inventory
   14.667 ERROR      python version = 3.6.8 (default, Jan 11 2019, 02:17:16) [GCC 8.2.1 20180905 (Red Hat 8.2.1-3)]
   14.667 ERROR    Using /etc/ansible/ansible.cfg as config file
   14.667 ERROR    setting up inventory plugins
   14.667 ERROR    Parsed /var/lib/awx/venv/awx/lib64/python3.6/site-packages/awx/plugins/inventory/vmware_inventory.py inventory source with script plugin**
   14.672 DEBUG    Finished loading from source: /var/lib/awx/venv/awx/lib64/python3.6/site-packages/awx/plugins/inventory/vmware_inventory.py
   14.672 INFO     Processing JSON output...
   14.672 DEBUG    Loaded group: all
   14.672 DEBUG    Loaded group: None
   14.672 DEBUG    Adding child group None to parent all
   14.672 DEBUG    Loaded host: cmnv-fs01_564d019c-c423-d8dc-735b-c3d379606541
   14.672 DEBUG    Adding host cmnv-fs01_564d019c-c423-d8dc-735b-c3d379606541 to group None
   14.672 DEBUG    Adding child group None to parent all
   14.672 DEBUG    Adding child group None to parent all
   14.672 DEBUG    Loaded group: centos64Guest
   14.672 DEBUG    Adding child group centos64Guest to parent all
   14.672 DEBUG    Loaded group: centos6_64Guest
   14.672 DEBUG    Adding child group centos6_64Guest to parent all
   14.672 DEBUG    Loaded group: centos7_64Guest
   14.673 DEBUG    Adding child group centos7_64Guest to parent all
   14.673 DEBUG    Loaded group: debian9_64Guest
   14.673 DEBUG    Adding child group debian9_64Guest to parent all
   14.673 DEBUG    Loaded group: other26xLinux64Guest
   14.673 DEBUG    Adding child group other26xLinux64Guest to parent all
   14.673 DEBUG    Loaded group: other3xLinux64Guest
   14.673 DEBUG    Adding child group other3xLinux64Guest to parent all
   14.673 DEBUG    Loaded group: rhel4_64Guest
   14.673 DEBUG    Adding child group rhel4_64Guest to parent all
   14.673 DEBUG    Loaded group: rhel7_64Guest
   14.673 DEBUG    Adding child group rhel7_64Guest to parent all
   14.673 DEBUG    Loaded group: sles11_64Guest
   14.673 DEBUG    Adding child group sles11_64Guest to parent all
   14.673 DEBUG    Loaded group: windows8Server64Guest
   14.673 DEBUG    Adding child group windows8Server64Guest to parent all
   14.673 DEBUG    Loaded group: windows9Server64Guest
   14.673 DEBUG    Adding child group windows9Server64Guest to parent all
   14.673 DEBUG    Loaded host: cmnv-mgmt3fom_564d8421-8256-ab97-f994-a3694b881991
   14.673 DEBUG    Adding host cmnv-mgmt3fom_564d8421-8256-ab97-f994-a3694b881991 to group centos64Guest
   14.673 DEBUG    Loaded host: cmnv-mgmtvsa1_564ded04-94e6-6508-94b9-b4e05f926bc5
   14.673 DEBUG    Adding host cmnv-mgmtvsa1_564ded04-94e6-6508-94b9-b4e05f926bc5 to group centos64Guest
   14.673 DEBUG    Loaded host: cmnv-mgmtvsa2_564de7e5-3e85-9e20-d1b5-c7e38d732247
   14.673 DEBUG    Adding host cmnv-mgmtvsa2_564de7e5-3e85-9e20-d1b5-c7e38d732247 to group centos64Guest
   14.673 DEBUG    Loaded host: cmnv-storvsa01_564d80d3-f4f6-3bfb-6e7d-667cd7ff64ac
   14.674 DEBUG    Adding host cmnv-storvsa01_564d80d3-f4f6-3bfb-6e7d-667cd7ff64ac to group centos64Guest
   14.674 DEBUG    Loaded host: cmnv-storvsa02_564d78f3-0abc-de3a-c5f6-7d4f13321d83
   14.674 DEBUG    Adding host cmnv-storvsa02_564d78f3-0abc-de3a-c5f6-7d4f13321d83 to group centos64Guest
   14.674 DEBUG    Loaded host: cmnv-storvsa03_564d23c6-7d56-518f-341f-528a49c0750d
   14.674 DEBUG    Adding host cmnv-storvsa03_564d23c6-7d56-518f-341f-528a49c0750d to group centos64Guest
   14.674 DEBUG    Loaded host: cmnv-storvsa04_564df456-8c00-696f-4337-9b82026a4412
   14.674 DEBUG    Adding host cmnv-storvsa04_564df456-8c00-696f-4337-9b82026a4412 to group centos64Guest
   14.674 DEBUG    Adding child group centos64Guest to parent all
   14.674 DEBUG    Loaded host: cmrv-ov4vc_423d58d8-dd53-1b48-0c21-c22a2c1b07b8
   14.674 DEBUG    Adding host cmrv-ov4vc_423d58d8-dd53-1b48-0c21-c22a2c1b07b8 to group centos6_64Guest
   14.674 DEBUG    Adding child group centos6_64Guest to parent all
   14.674 DEBUG    Loaded host: crnv-cohesity_423ddc5c-ae27-cbc4-8806-0c7df069ef14
   14.674 DEBUG    Adding host crnv-cohesity_423ddc5c-ae27-cbc4-8806-0c7df069ef14 to group centos7_64Guest
   14.674 DEBUG    Adding child group centos7_64Guest to parent all
   14.674 DEBUG    Loaded host: cmrv-ssmc01_423de3e3-541a-12d1-3d76-bb62d9c82f4a
   14.674 DEBUG    Adding host cmrv-ssmc01_423de3e3-541a-12d1-3d76-bb62d9c82f4a to group debian9_64Guest
   14.674 DEBUG    Loaded host: crnv-iloamplifier01_423dfa83-deed-faab-cf1d-f7746df04814
   14.674 DEBUG    Adding host crnv-iloamplifier01_423dfa83-deed-faab-cf1d-f7746df04814 to group debian9_64Guest
   14.674 DEBUG    Adding child group debian9_64Guest to parent all
   14.674 DEBUG    Loaded host: cmrv-globalov01_423db291-f709-83d4-d316-a05fa59f852f
   14.674 DEBUG    Adding host cmrv-globalov01_423db291-f709-83d4-d316-a05fa59f852f to group other26xLinux64Guest
   14.674 DEBUG    Loaded host: cmrv-mgmtov4_4206a368-511d-90b6-d29b-0d69417a169b
   14.674 DEBUG    Adding host cmrv-mgmtov4_4206a368-511d-90b6-d29b-0d69417a169b to group other26xLinux64Guest
   14.674 DEBUG    Loaded host: cmrv-ovsvr01_564d9ad9-4fe1-1654-9837-d0f505b2a5f6
   14.674 DEBUG    Adding host cmrv-ovsvr01_564d9ad9-4fe1-1654-9837-d0f505b2a5f6 to group other26xLinux64Guest
   14.674 DEBUG    Loaded host: cmvr-objectov02_423d0fd3-b912-eed9-4b94-4a11d54c209c
   14.675 DEBUG    Adding host cmvr-objectov02_423d0fd3-b912-eed9-4b94-4a11d54c209c to group other26xLinux64Guest
   14.675 DEBUG    Adding child group other26xLinux64Guest to parent all
   14.675 DEBUG    Loaded host: cmrv-3par01vsp_423d16be-34a8-135c-8735-278329eac301
   14.675 DEBUG    Adding host cmrv-3par01vsp_423d16be-34a8-135c-8735-278329eac301 to group other3xLinux64Guest
   14.675 DEBUG    Loaded host: cmrv-vc01_564d7656-47a6-56dd-78bc-55e9751f1079
   14.675 DEBUG    Adding host cmrv-vc01_564d7656-47a6-56dd-78bc-55e9751f1079 to group other3xLinux64Guest
   14.675 DEBUG    Adding child group other3xLinux64Guest to parent all
   14.675 DEBUG    Loaded host: bootstrap-0_423d81c6-74d1-bd48-f1a1-6edd2d7958d3
   14.675 DEBUG    Adding host bootstrap-0_423d81c6-74d1-bd48-f1a1-6edd2d7958d3 to group rhel4_64Guest
   14.675 DEBUG    Loaded host: compute-0_423dd553-fe9e-26ba-0422-cd2b9479e02e
   14.675 DEBUG    Adding host compute-0_423dd553-fe9e-26ba-0422-cd2b9479e02e to group rhel4_64Guest
   14.675 DEBUG    Loaded host: compute-1_423d3ba6-8f9e-100e-8b23-9a00809dd1ac
   14.675 DEBUG    Adding host compute-1_423d3ba6-8f9e-100e-8b23-9a00809dd1ac to group rhel4_64Guest
   14.675 DEBUG    Loaded host: compute-2_423d9957-57e2-2257-917c-27658d54d2da
   14.675 DEBUG    Adding host compute-2_423d9957-57e2-2257-917c-27658d54d2da to group rhel4_64Guest
   14.675 DEBUG    Loaded host: control-plane-0_423dc264-f64b-1893-cbef-0a13d0bb6279
   14.675 DEBUG    Adding host control-plane-0_423dc264-f64b-1893-cbef-0a13d0bb6279 to group rhel4_64Guest
   14.675 DEBUG    Loaded host: control-plane-1_423d7b12-25c7-539d-85a2-4d5a5e62ad1c
   14.675 DEBUG    Adding host control-plane-1_423d7b12-25c7-539d-85a2-4d5a5e62ad1c to group rhel4_64Guest
   14.675 DEBUG    Loaded host: control-plane-2_423db1be-ca50-d0ab-c194-108eab9e3c4d
   14.675 DEBUG    Adding host control-plane-2_423db1be-ca50-d0ab-c194-108eab9e3c4d to group rhel4_64Guest
   14.675 DEBUG    Adding child group rhel4_64Guest to parent all
   14.675 DEBUG    Loaded host: cmrv-rhans01_423debe7-3315-cd75-240d-2676219b9466
   14.675 DEBUG    Adding host cmrv-rhans01_423debe7-3315-cd75-240d-2676219b9466 to group rhel7_64Guest
   14.675 DEBUG    Loaded host: cmrv-rhsat01_564d6796-81c3-ddc0-3d51-bb9f2a37175c
   14.675 DEBUG    Adding host cmrv-rhsat01_564d6796-81c3-ddc0-3d51-bb9f2a37175c to group rhel7_64Guest
   14.675 DEBUG    Adding child group rhel7_64Guest to parent all
   14.676 DEBUG    Loaded host: cmrv-vro75_423d348a-1634-a8c4-a49a-e4c281c8da1b
   14.676 DEBUG    Adding host cmrv-vro75_423d348a-1634-a8c4-a49a-e4c281c8da1b to group sles11_64Guest
   14.676 DEBUG    Loaded host: cmrv-vrops01_423de9a0-e085-4f6c-7a88-ca97c538c882
   14.676 DEBUG    Adding host cmrv-vrops01_423de9a0-e085-4f6c-7a88-ca97c538c882 to group sles11_64Guest
   14.676 DEBUG    Adding child group sles11_64Guest to parent all
   14.676 DEBUG    Loaded host: cmnv-tower3_423dcab1-25fc-a14d-1ec8-a35df1334871
   14.676 DEBUG    Adding host cmnv-tower3_423dcab1-25fc-a14d-1ec8-a35df1334871 to group all
   14.676 DEBUG    Loaded host: cmrv-ssmcbna02_423d5138-6aa7-d328-7e7e-09669436e9b5
   14.676 DEBUG    Adding host cmrv-ssmcbna02_423d5138-6aa7-d328-7e7e-09669436e9b5 to group windows8Server64Guest
   14.676 DEBUG    Adding child group windows8Server64Guest to parent all
   14.676 DEBUG    Loaded host: WINTEST01_423d47e7-699d-fa56-aa47-101c54cd827f
   14.676 DEBUG    Adding host WINTEST01_423d47e7-699d-fa56-aa47-101c54cd827f to group windows9Server64Guest
   14.676 DEBUG    Loaded host: Windows 2016 Template_423d93a6-6529-f39b-9d46-73728babec8d
   14.676 DEBUG    Adding host Windows 2016 Template_423d93a6-6529-f39b-9d46-73728babec8d to group windows9Server64Guest
   14.676 DEBUG    Loaded host: cmnv-sndbox1_423df03f-5272-a4f2-b4ff-6cc93c34b6df
   14.678 DEBUG    Adding host cmnv-sndbox1_423df03f-5272-a4f2-b4ff-6cc93c34b6df to group windows9Server64Guest
   14.678 DEBUG    Loaded host: cmrv-dc01_4228f3d7-6bc8-61e7-4b3d-52de35130a3e
   14.678 DEBUG    Adding host cmrv-dc01_4228f3d7-6bc8-61e7-4b3d-52de35130a3e to group windows9Server64Guest
   14.678 DEBUG    Loaded host: cmrv-dc02_42280e73-f43b-87a1-a797-5496e911862a
   14.678 DEBUG    Adding host cmrv-dc02_42280e73-f43b-87a1-a797-5496e911862a to group windows9Server64Guest
   14.678 DEBUG    Loaded host: cmrv-jumpsvr1_421cce24-fc6b-c7a3-bf29-46ebd5bd84c7
   14.678 DEBUG    Adding host cmrv-jumpsvr1_421cce24-fc6b-c7a3-bf29-46ebd5bd84c7 to group windows9Server64Guest
   14.678 DEBUG    Loaded host: cmrv-jumpsvr2_564d8834-1a4e-f65f-886e-09c837e03e74
   14.678 DEBUG    Adding host cmrv-jumpsvr2_564d8834-1a4e-f65f-886e-09c837e03e74 to group windows9Server64Guest
   14.678 DEBUG    Loaded host: cmrv-logmon01_423d2861-4e9d-ee1b-0bdb-be4b5a08fec9
   14.678 DEBUG    Adding host cmrv-logmon01_423d2861-4e9d-ee1b-0bdb-be4b5a08fec9 to group windows9Server64Guest
   14.678 DEBUG    Loaded host: cmrv-mgmt01_421c2f8f-f5c4-4c41-3534-cec059f4515a
   14.678 DEBUG    Adding host cmrv-mgmt01_421c2f8f-f5c4-4c41-3534-cec059f4515a to group windows9Server64Guest
   14.678 DEBUG    Loaded host: cmrv-snmid01_564d7701-074d-3e42-5879-25f934b5b682
   14.678 DEBUG    Adding host cmrv-snmid01_564d7701-074d-3e42-5879-25f934b5b682 to group windows9Server64Guest
   14.678 DEBUG    Loaded host: cmrv-util01_422815d6-5ab2-ea41-f14a-b6486f808668
   14.678 DEBUG    Adding host cmrv-util01_422815d6-5ab2-ea41-f14a-b6486f808668 to group windows9Server64Guest
   14.678 DEBUG    Loaded host: cmrv-vbrsvr01_423d7694-7add-2dc6-c2ff-228afc50ad3e
   14.678 DEBUG    Adding host cmrv-vbrsvr01_423d7694-7add-2dc6-c2ff-228afc50ad3e to group windows9Server64Guest
   14.678 DEBUG    Loaded host: ucpny-mgmt01_423d8ed9-30ae-56f3-4559-65537ef2d46c
   14.678 DEBUG    Adding host ucpny-mgmt01_423d8ed9-30ae-56f3-4559-65537ef2d46c to group windows9Server64Guest
   14.678 DEBUG    Adding child group windows9Server64Guest to parent all
   14.679 INFO     Loaded 12 groups, 44 hosts
   14.694 DEBUG    Inventory variables unmodified
   14.702 DEBUG    Group "None" added
   14.706 DEBUG    Group "centos64Guest" added
   14.710 DEBUG    Group "centos6_64Guest" added
.......

**Comment by kdelee** [comment]
@sankat447 thanks for trying this out. That error is just ansible being loud, it is not actually an error (wish we could get it to look less scary, but that is a separate issue)

**Comment by kdelee** [comment]
https://github.com/ansible/awx/commit/008fe42b4deb4bad93699142f21083e7b8ec0161 updated the script and it has been working for me in python 2 and 3. Going to close this, I encourage anyone that has participated in this thread to try this out and let me know if they still have problems.
